src/qutip_qip/vqa.py

```diff
@@ -0,0 +1,842 @@
+        else:
+            if angles is None:
+                raise ValueError("No parameter given")
+
+        if isinstance(self.operator, ParameterizedHamiltonian):
+            return (-1j * self.operator.get_hamiltonian(angles)).expm()
+        else:
+            if len(angles) != 1:
+                raise ValueError(
+                    "Expected one angle for singly-parameterized"
+                    " Hamiltonian."
+                )
+            else:
+                return (-1j * angles[0] * self.operator).expm()
+
+    def get_unitary_derivative(self, angles):
+        """
+        Compute the derivative of the block's unitary with respect to its
+        free parameter, assuming it is of the form e^{-i * theta * H}
+        for a free parameter theta.
+
+        Parameters
+        ----------
+        angle: list of float
+            free parameters to take derivatives with respect to
+
+        Returns
+        -------
+        derivative: float
+        """
+        if self.is_unitary or self.is_native_gate:
+            raise ValueError(
+                "Can only take derivative of block specified "
+                "by Hamiltonians or ParameterizedHamiltonian instances."
+            )
+        if isinstance(self.operator, ParameterizedHamiltonian):
+            return self.get_unitary_frechet_derivative(angles)
+        else:
+            if len(angles) != 1:
+                raise ValueError(
+                    "Expected a single angle for non-"
+                    "ParameterizedHamiltonian instance."
+                )
+            return self.get_unitary(angles) * -1j * self.operator
+
+    def get_unitary_frechet_derivative(self, angles, term_index=0):
+        """
+        Compute the derivative of the block's unitary with respect
+        to a free parameter, using the Frechet derivative of
+        the exponential function.
+
+        Parameters
+        ----------
+        angles: list of float
+            the free parameters to take derivatives with respect to
+        term_index: int, optional
+            Index of Parameterized Hamiltonian term that specifies the matrix
+            direction in which to take the derivative.
+
+        Returns
+        -------
+        derivative: float
+        """
+        if self.is_unitary or self.is_native_gate:
+            raise ValueError(
+                "Can only take frechet derivative of block "
+                "specified by Hamiltonians"
+            )
+        if isinstance(self.operator, ParameterizedHamiltonian):
+            arg = -1j * self.operator.get_hamiltonian(angles)
+            direction = -1j * self.operator.p_terms[term_index]
+        else:
+            if len(angles) != 1:
+                raise ValueError(
+                    "Expected a single angle for non-"
+                    "ParameterizedHamiltonian instance."
+                )
+            arg = -1j * self.operator * angles[0]
+            direction = -1j * self.operator
```

Comment:
This part has some overlap with the `get_unitary_derivative` method. The method `get_unitary_frechet_derivative` is only used if the operator is `ParameterizedHamiltonian`.